File: scripts/demo-websocket.py
#!/usr/bin/env python3
"""
WebSocket 实时推送演示脚本

流程:
1. 打开 WebSocket 演示页面 (websocket-demo.html)
2. 等待 WebSocket 连接建立
3. 将已有订单重置为待支付状态 (status=1, pay_status=0)
4. 调用支付确认接口 /notify/mockPaySuccess?orderNumber=xxx
5. 后端 paySuccess() 发送 WebSocket 通知 → 演示页显示消息
6. 截图保存结果
"""

import subprocess
import time
import json
from playwright.sync_api import sync_playwright

MYSQL_CRED = "-uroot -p123456"
BUSINESS_GATEWAY = "http://localhost:8090"

def reset_order_for_demo(order_number):
    """将已有订单重置为待支付状态"""
    cmds = [
        f'mysql {MYSQL_CRED} -e "USE lgg_ruoyi; UPDATE lgg_orders SET status=1, pay_status=0 WHERE number={order_number} AND status=2;" 2>&1',
        f'mysql {MYSQL_CRED} -e "USE lgg_ruoyi; SELECT id, number, status, pay_status FROM lgg_orders WHERE number={order_number};" 2>&1'
    ]
    for cmd in cmds:
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        print(f"[DB] {result.stdout.strip()}")
        if result.stderr.strip():
            print(f"[DB-ERR] {result.stderr.strip()}")

def trigger_payment(order_number):
    """调用支付确认接口"""
    url = f"{BUSINESS_GATEWAY}/notify/mockPaySuccess?orderNumber={order_number}"
    print(f"[API] Calling: {url}")
    result = subprocess.run(
        ["curl", "-s", "-o", "/dev/null", "-w", "%{http_code}", url],
        capture_output=True, text=True, timeout=10
    )
    print(f"[API] Response code: {result.stdout}")
    return result.stdout

def verify_order_status(order_number):
    """验证订单状态已更新"""
    cmd = f'mysql {MYSQL_CRED} -e "USE lgg_ruoyi; SELECT id, number, status, pay_status FROM lgg_orders WHERE number={order_number};" 2>&1'
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    print(f"[DB-VERIFY] {result.stdout.strip()}")
    return result.stdout

def main():
    order_number = "1781093884077"  # Order #72, was status=2 (TO_BE_CONFIRMED)

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1280, "height": 800}
        )
        page = context.new_page()

        # 1. Open WebSocket demo page
        html_path = "file:///Users/caolei/Desktop/springboot-lgg/websocket-demo.html"
        print(f"[BROWSER] Opening {html_path}")
        page.goto(html_path, wait_until="networkidle")
        time.sleep(2)

        # Take initial screenshot
        page.screenshot(path="/tmp/ws-demo-01-initial.png", full_page=True)
        print("[SCREENSHOT] ws-demo-01-initial.png saved")

        # 2. Wait for WebSocket connection (check for the "connected" log entry)
        print("[WS] Waiting for WebSocket connection...")
        try:
            page.wait_for_selector("text=已连接", timeout=15000)
            print("[WS] Connected!")
        except Exception as e:
            print(f"[WS] Connection timeout: {e}")

        time.sleep(1)
        page.screenshot(path="/tmp/ws-demo-02-connected.png", full_page=True)
        print("[SCREENSHOT] ws-demo-02-connected.png saved")

        # 3. Reset order to pending payment
        print(f"[DEMO] Resetting order {order_number} to PENDING_PAYMENT...")
        reset_order_for_demo(order_number)
        time.sleep(1)

        # 4. Trigger payment (this sends WebSocket notification)
        print(f"[DEMO] Triggering paySuccess for order {order_number}...")
        trigger_payment(order_number)
        time.sleep(2)

        # 5. Wait for WebSocket message to appear
        print("[WS] Waiting for WebSocket message...")
        try:
            page.wait_for_selector("text=来单提醒", timeout=10000)
            print("[WS] Notification received!")
        except Exception as e:
            print(f"[WS] Notification timeout: {e}")

        # 6. Take final screenshot
        time.sleep(1)
        page.screenshot(path="/tmp/ws-demo-03-notification.png", full_page=True)
        print("[SCREENSHOT] ws-demo-03-notification.png saved")

        # 7. Verify DB state
        print("[DEMO] Verifying order status after payment...")
        verify_order_status(order_number)

        # Wait a moment for user to see
        time.sleep(2)

        browser.close()
        print("[DONE] Demo completed! Screenshots at /tmp/ws-demo-*.png")

if __name__ == "__main__":
    main()

File: scripts/demo-websocket.js
const puppeteer = require('puppeteer');
const { execSync } = require('child_process');
const path = require('path');

const http = require('http');
const fs = require('fs');

const ORDER_NUMBER = '1781093884077';
const DB_CRED = '-uroot -p123456';
const GATEWAY = 'http://localhost:8090';
const HTML_DIR = '/Users/caolei/Desktop/springboot-lgg';
const SCREENSHOT_DIR = '/tmp/ws-demo';

function run(cmd) {
  return execSync(cmd, { shell: true, timeout: 15000, encoding: 'utf-8' }).trim();
}

function serveHtml(port) {
  const server = http.createServer((req, res) => {
    const content = fs.readFileSync(HTML_DIR + '/websocket-demo.html', 'utf-8');
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(content);
  });
  server.listen(port);
  return server;
}

(async () => {
  fs.mkdirSync(SCREENSHOT_DIR, { recursive: true });
  const HTTP_PORT = 18765;
  const server = serveHtml(HTTP_PORT);
  const HTML_URL = `http://localhost:${HTTP_PORT}/`;
  console.log(`HTTP server on port ${HTTP_PORT}`);
  console.log('=== WebSocket Real-Time Push Demo ===\n');

  // 1. Reset order to PENDING_PAYMENT with fresh order_time to avoid auto-cancel
  console.log('[1/5] Resetting order #72 to PENDING_PAYMENT...');
  run(`mysql ${DB_CRED} -e "USE lgg_ruoyi; UPDATE lgg_orders SET status=1, pay_status=0, order_time=NOW() WHERE id=72;"`);
  const before = run(`mysql ${DB_CRED} -e "USE lgg_ruoyi; SELECT id, number, status, pay_status, order_time FROM lgg_orders WHERE id=72;" -t`);
  console.log(before + '\n');

  // 2. Launch system Chrome via puppeteer
  console.log('[2/5] Launching system Chrome...');
  const browser = await puppeteer.launch({
    headless: true,
    executablePath: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  const page = await browser.newPage();
  await page.setViewport({ width: 1280, height: 800 });

  // Listen to console messages from the page
  page.on('console', msg => console.log(`  [PAGE ${msg.type()}] ${msg.text()}`));

  // 3. Open WebSocket demo page
  console.log(`[3/5] Opening ${HTML_URL}`);
  await page.goto(HTML_URL, { waitUntil: 'load', timeout: 15000 });
  await new Promise(r => setTimeout(r, 3000));

  // Screenshot: after connection
  await page.screenshot({ path: path.join(SCREENSHOT_DIR, '01-connected.png'), fullPage: true });
  const text1 = await page.evaluate(() => document.body.innerText);
  console.log(`  Screenshot 1 saved`);
  console.log(`  Page: ${text1.substring(0, 200).replace(/\n/g, ' | ')}...\n`);

  // 4. Trigger payment
  console.log('[4/5] Triggering paySuccess...');
  const httpCode = run(`curl -s '${GATEWAY}/notify/mockPaySuccess?orderNumber=${ORDER_NUMBER}' --noproxy '*' -o /dev/null -w "%{http_code}"`);
  console.log(`  HTTP Status: ${httpCode}`);

  // Wait for WebSocket notification to arrive
  await new Promise(r => setTimeout(r, 3000));

  // Screenshot: after notification
  await page.screenshot({ path: path.join(SCREENSHOT_DIR, '02-notification.png'), fullPage: true });
  const text2 = await page.evaluate(() => document.body.innerText);
  console.log(`  Screenshot 2 saved`);
  console.log(`  Page: ${text2.replace(/\n/g, ' | ')}\n`);

  // 5. Verify DB
  console.log('[5/5] Verifying order status...');
  const after = run(`mysql ${DB_CRED} -e "USE lgg_ruoyi; SELECT id, number, status, pay_status FROM lgg_orders WHERE id=72;" -t`);
  console.log(after + '\n');

  await browser.close();
  server.close();
  console.log('=== Demo Complete ===');
  console.log(`Screenshots: ${SCREENSHOT_DIR}/`);
  console.log(`Open with: open ${SCREENSHOT_DIR}/`);
})();
